### PR TITLE
make_parallel for network containers

### LINKED ISSUE #762: Implement make_parallel for network containers
So we can have parallel version for any network built using Sequential, Parallel and Branch.

### PR DESCRIPTION
**Pull request by emailweixu**
This change supports creating parallel network for network containers.  This is achieved by implementing make_parallel for various layers and transforming all the modules in the container.

With this change, it will be trivial to implement make_parallel for all kinds of networks if they are implemented using network containers.

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -2854,3 +2912,146 @@ def reset_parameters(self):
...
+    Examples:
+
+    Applying parallel net on same input:
+
+    .. code-block:: python
+
+        pnet = make_parallel_net(net, n)
+        # replicate input.
+        # pinput will have shape [batch_size, n, ...], if input has shape [batch_size, ...]
+        pinput = make_parallel_input(input)
```
Comment (le-horizon) [inline alf/layers.py]: `make_parallel_input(input, n)`

File: alf/layers.py [hunk 2]
```diff
@@ -13,14 +13,14 @@
...
+from absl import logging
 import copy
-
 from functools import partial
 import numpy as np
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
-from typing import Callable
+from typing import Optional, Union, Callable
```
Comment (Haichao-Zhang) [inline alf/layers.py]: `Optional` is not used

File: alf/layers.py [hunk 3]
```diff
@@ -2854,3 +2912,146 @@ def reset_parameters(self):
...
+        return NaiveParallelLayer(module, n)
+
+
+class NaiveParallelLayer(nn.Module):
+    def __init__(self, module: Union[nn.Module, Callable], n: int):
+        """
+        A parallel network has ``n`` copies of network with the same structure but
+        different indepently initialized parameters.
+
+        ``NaiveParallelLayer`` creats ``n`` independent networks with the same
```
Comment (Haichao-Zhang) [inline alf/layers.py]: creats -> creates

File: alf/layers_test.py [hunk 4]
```diff
@@ -932,6 +963,82 @@ def test_sequential4(self):
...
+        self._test_make_parallel(layer, input_spec)
+        layer = alf.layers.Transpose(-2, -1)
+        self._test_make_parallel(layer, input_spec)
+
+    def test_permute(self):
+        input_spec = alf.BoundedTensorSpec((8, 4, 10))
+        layer = alf.layers.Permute(2, 1, 0)
+        self._test_make_parallel(layer, input_spec)
+
+    def test_onehost(self):
```
Comment (Haichao-Zhang) [inline alf/layers_test.py]: test_onehost -> test_onehot

File: alf/networks/networks.py [hunk 5]
```diff
@@ -171,7 +178,8 @@ def __init__(self,
                  stack_size,
                  pooling_size=1,
                  dtype=torch.float32,
-                 mode='skip'):
+                 mode='skip',
+                 name='TemporalPool'):
```
Comment (le-horizon) [inline alf/networks/networks.py:182]: Looking at the TemporalPool example above, it seems if pool size == 2, we always ignore every other (half) of the timesteps.
Have we considered pooling every timestep, instead of pooling every pool_size timesteps?

**Review by le-horizon — COMMENTED** [review]
Thanks for the nice improvement Wei.

Looks good as is, just small comments.

Thanks,
Le

File: alf/layers.py [hunk 6]
```diff
@@ -2624,6 +2664,9 @@ def __init__(self, dim):
     def forward(self, input):
         return input.sum(dim=self._dim)
 
+    def make_parallel(self, n: int):
+        return Sum(self._dim)
```
Comment (Haichao-Zhang) [inline alf/layers.py:2647]: Might be useful to add comments on the assumed shape of parallel batch `[B, n, ...]`, and both the batch dimension and replica dimension are not counted for dim.

File: alf/layers.py [hunk 7]
```diff
@@ -109,6 +112,9 @@ def __init__(self, dim0=0, dim1=1):
     def forward(self, x):
         return x.transpose(self._dim0, self._dim1)
 
+    def make_parallel(self, n: int):
```
Comment (Haichao-Zhang) [inline alf/layers.py:116]: Would be useful to add comments on the assumed shape of parallel batch `[B, n, ...]`, and both the batch dimension and replica dimension are not considered for transpose.

File: alf/layers.py (hunk 6 above)
Comment (hnyu) [reply alf/layers.py:2647]: > Might be useful to add comments on the assumed shape of parallel batch `[B, n, ...]`, and both the batch dimension and replica dimension are not counted for dim.

If the replica dim is not counted, then the value of `self._dim` is not correct here.

File: alf/layers.py [hunk 8]
```diff
@@ -2854,3 +2912,146 @@ def reset_parameters(self):
...
+        return Sequential(
+            *new_networks, output=self._output, **new_named_networks)
+
+
+def make_parallel_net(module, n: int):
+    """Make a parallelized version of ``module``.
+
+    A parallel network has ``n`` copies of network with the same structure but
+    different independently initialized parameters. The parallel network can
+    process a batch of the data with shape [batch_size, n, ...] using ``n``
```
Comment (hnyu) [inline alf/layers.py:2921]: Should we automatically convert data with shape [batch_size, ...] to [batch_size, n, ...]?

File: alf/nest/utils.py [hunk 9]
```diff
@@ -104,6 +105,10 @@ def _combine_flat(self, tensors):
         else:
             return torch.cat(tensors, dim=self._dim)
 
+    def make_parallel(self, n):
+        dim = self._dim if self._dim < 0 else self._dim + 1
```
Comment (hnyu) [inline alf/nest/utils.py]: Should add comments saying that self._dim excludes batch dim and parallel dim. The current implementation seems not to ignore the batch dim.

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/layers.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/layers.py:2647]: Base on my understanding, the replica dim is excluded. In this case the ``self._dim`` is correct as it is increased by `1` during the construction of the new``Sum`` instance. The same is for `Transpose` and several other related operators. Some more documentation on this would be helpful.

File: alf/layers.py (hunk 8 above)
Comment (emailweixu) [reply alf/layers.py:2921]: As network can have sub-networks, doing this check for all of them can be wasteful. So it's intended to use make_parallel_input to do the conversion by the user.

File: alf/nest/utils.py (hunk 9 above)
Comment (emailweixu) [reply alf/nest/utils.py]: Changed the behavior of NestConcat to not including batch dim.
Comments added.

File: alf/networks/networks.py (hunk 5 above)
Comment (emailweixu) [reply alf/networks/networks.py:182]: Is mode "avg" what you want?

File: alf/layers.py [hunk 10]
```diff
@@ -2665,6 +2634,16 @@ def forward(self, input):
         return input.sum(dim=self._dim)
 
     def make_parallel(self, n: int):
+        """Create a Sum layer to handle parallel batch.
+
+        It is assumed that a parallel batch has shape [B, n, ...] and both the
+        batch dimension and replica dimension are counted for ``dim``
```
Comment (Haichao-Zhang) [inline alf/layers.py]: `are counted` --> `are not counted`

**Review by Haichao-Zhang — COMMENTED** [review]
The comments are very helpful. Just some more minor points.

File: alf/utils/spec_utils.py [hunk 11]
```diff
@@ -17,13 +17,57 @@
...
 from typing import Iterable
 
-from alf.layers import BatchSquash
 import alf.nest as nest
 from alf.nest.utils import get_outer_rank
 from alf.tensor_specs import TensorSpec, BoundedTensorSpec
 from . import dist_utils
 
 
+class BatchSquash(object):
```
Comment (hnyu) [inline alf/utils/spec_utils.py]: It's better to put this in tensor_utils?

**Review by Haichao-Zhang — APPROVED** [review]

**Review by hnyu — APPROVED** [review]
